README.md
---
page_type: meta
languages:
- python
description: "README"
urlFragment: readme
---
# Elo Football Data Calculator

## Workflow

### Local Development

```shell
    > source init.sh                   # Starts virtual environment and sets config variables
    > pip install -r requirements.txt  # Installs Python dependencies
    > npm install                      # installs NPM dependencies
    > npm run update:all               # Updates Elo HDF5 databases
    > npm run build                    # Builds static assets
    > npm run serve                    # Starts configured development server
```

### Deployment

There are 2 git remotes, and 2 corresponding heroku apps. Push to `stage` for staging and `pro` for production.
The app is hosted on a [gunicorn](https://gunicorn.org/) server.

```bash
    > git push stage master vX.X.X
    > git push pro master vX.X.X
```

## Code Structure

### Logic

The mathematical logic is contained in `elo.py`.

### Api Request Handling

`FBApi.py` serves as the dedicated api request handler.

### Construction of Data Objects

The data objects are stored in a database as HD5 tables and read by pandas. They are written by `updater.py`.

## Documentation

## API Requests

Complete [API documentation](https://www.football-data.org/documentation/api). API requires signing up for an account
to recieve an API key.

## League Coverage

| id   | code | name                  | area        |
|------|------|-----------------------|-------------|
| 2002 | BL   | Bundesliga            | Germany     |
| 2015 | FL1  | Ligue 1               | France      |
| 2021 | PL   | Premier League        | England     |
| 2016 | ELC  | Championship          | England     |
| 2014 | PD   | La Liga               | Spain       |
| 2019 | SA   | Serie A               | Italy       |
| 2017 | PPL  | Primeira Liga         | Portugal    |
| 2003 | DED  | Eredivisie            | Netherlands |


CHANGELOG.md
---
page_type: meta
languages:
- python
description: "CHANGELOG"
urlFragment: changelog
---
# Elo Football App

## Release Notes

Changelog labels are any of the following:

- [BUGFIX], [CHANGE], [CODEQUALITY], [DEPENDENCY], [ENHANCEMENT], [OTHER], [TRANSLATION]

### Version 1.0.1 (2021-03-02)

- [DEPENDENCY] Updated dependencies.

### Version 1.0.0 (2021-01-12)

#### Major Release 1.0

- [CODEQUALITY] Added command line directives for updating databases.
- [CODEQUALITY] Added a GitHub Actions workflow to lint python on `git push`
- [DEPENDENCY] Upgraded all dependencies.

### Version 0.9.3 (2021-01-06)

- [BUGFIX] Reverted previous change that ran the competitions updater persistently (400 errors).

### Version 0.9.2 (2021-01-02)

- [ENHANCEMENT] Added area flag SVG and current matchday to card titles.
- [CODEQUALITY] `npm run update` now also updates the `\comps` table for current matchday info.

### Version 0.9.1 (2020-12-30)

- [CHANGE] Switched to backend table sortings (pandas), sacrificing speed for greater flexibility
- [ENHANCEMENT] Some cosmetic UI changes

### Version 0.9.0 (2020-12-23)

#### Minor Release 0.9

- [OTHER] Successful deployment to heroku with Python 3.9.1.

### Version 0.8.0 (2020-11-17)

#### Minor Release 0.8

- [ENHANCEMENT] Added Teams tab, for future graphing capabilities.

### Version 0.7.0 (2020-11-13)

#### Minor Release 0.7

- [OTHER] Updated to 2020-2021 Season.
- [OTHER] New runtime: Python 3.9.0.

### Version 0.6.0 (2020-08-13)

#### Minor Release 0.6

- [ENHANCEMENT] Standings API incorporated into DataTable.

### Version 0.5.2 (2020-08-13)

- [CODEQUALITY] Updated 'comps' DataFrame to remove Pandas warning.
- [ENHANCEMENT] Customized app title and icon.

### Version 0.5.1 (2020-08-13)

- [DEPENDENCY] Updated UIKit to 3.5.6.

### Version 0.5.0 (2020-08-09)

#### Minor Release 0.5

- [BUGFIX] Bundesliga now working, all query filters are functional.
- [DEPENDENCY] Added PyTest, to implement in future release.
- [DEPENDENCY] Pandas upgraded to 1.1.0, throwing a deprecation warning — to keep in mind.
- [DEPENDENCY] Dash upgraded to 1.14. No significant effects.
- [OTHER] All 2019-2020 season Elo ratings finalized.
- [CODEQUALITY] Rework of class responsibilities; will be easier to expand to multi-year/multi-stage.

### Version 0.4.2 (2020-07-26)

- [OTHER] Updated runtime to Python 3.8.5.

### Version 0.4.1 (2020-07-26)

- [ENHANCEMENT] Added Elo Rank column to DataTable.
- [CODEQUALITY] Uninstalled unused Python packages.
- [DEPENDENCY] Updated to Plotly 4.9.

### Version 0.4.0 (2020-07-16)

#### Minor Release 0.4

- [ENHANCEMENT] Added last match results to data table.
- [ENHANCEMENT] Lots of UI improvements.

### Version 0.3.0 (2020-07-16)

#### Minor Release 0.3

- [ENHANCEMENT] Added custom UIKit theme and built from Sass.
- [DEPENDENCY] Grunt-contrib-sass, Ruby, Ruby-sass

### Version 0.2.1 (2020-07-15)

- [ENHANCEMENT] Edited conditional styles for readability and color contrast.
- [BUGFIX] Conditional styles and their legend now update with table selection.

### Version 0.2.0 (2020-07-15)

#### Minor Release 0.2

- [CODEQUALITY] Pandas backend now stores HDF5; no more JSON I/O.
- [CODEQUALITY] Grunt used to copy static web assets from UIKit.
- [DEPENDENCY] Grunt (dev) (NPM).
- [DEPENDENCY] UIKit upgraded to v3.5.5.

### Version 0.1.9 (2020-07-11)

- [ENHANCEMENT] Dash app is fully functional. 0.2 will release when integrated into heroku.

### Version 0.1.1 (2020-07-11)

- [CHANGE] Beginning transition to [Dash DataTable](https://dash.plotly.com/datatable) app.
- [DEPENDENCY] [Dash](https://dash.plotly.com), [Dash_DAQ](https://dash.plotly.com/dash-daq), [NumPy](https://numpy.org/doc/stable/reference/index.html), [Pandas](https://pandas.pydata.org/pandas-docs/stable/index.html)

### Version 0.1.0 (2020-07-11)

- [ENHANCEMENT] First working version with all competition routes and tables.
- [DEPENDENCY] [MPMath](http://mpmath.org/doc/current/)

### Version 0.0.3 (2020-07-03)

- [ENHANCEMENT] Now using [UIKit](https://getuikit.com/docs).
- [DEPENDENCY] UIKit (NPM)

### Version 0.0.2 (2020-07-03)

- [CHANGE] Migrated files from other reopsitory.

### Version 0.0.1 (2020-07-03)

- [ENHANCEMENT] [Deployed on Heroku](https://fs-mir-pro.herokuapp.com/).
- [CODEQUALITY] Changelog and SemVer initialized.


dashapp.py
import dash
import dash_core_components as dcc
import dash_html_components as html
import dash_daq as daq
import dash_table
from dash_table.Format import Format, Scheme
from dash.dependencies import Input, Output
import pandas as pd
import requests_cache


# Initialize requests cache
requests_cache.install_cache('fbd_cache', backend='sqlite', expire_after=86400)

# Load HDF5 Storage
_H5 = 'fbd_storage.h5'

# Competition selections
df_compsix = pd.read_hdf(_H5, 'comps')
df_compini = df_compsix[['label', 'value']]


def set_table_header(comp_code):
    comp_label = df_compsix.loc[comp_code]['label']
    comp_cs = df_compsix.loc[comp_code]['currentSeason']
    if df_compsix.loc[comp_code]['area'].get('ensignUrl'):
        area_logo = html.Img(
            style={'height': '32px'},
            src=df_compsix.loc[comp_code]['area']['ensignUrl']
        )
    else:
        area_logo = html.Div()
    comp_header = ' '.join([
        comp_label, chr(8211),
        'Matchday:',
        str(comp_cs['currentMatchday'])
    ])
    full_title = [
        area_logo,
        html.Span(
            comp_header,
            className='uk-padding-small'
        )
    ]
    return full_title


initial_comp = 'PL'
card_title = set_table_header(initial_comp)


def comp_elo_df(comp_code):
    hdf_key = comp_code.lower()
    df_static = pd.read_hdf(_H5, hdf_key)
    df_static['lastFix'] = df_static['fixtures'].apply(lambda x: df_static.loc[x[-1]]['tla'])
    res_map = {0: 'L', 0.5: 'D', 1: 'W'}
    df_static['lastRes'] = df_static['results'].apply(lambda x: res_map[x[-1]])
    df_static['lastDiff'] = df_static['eloRun'].apply(lambda x: x[-1] - x[-2])
    df_static['eloRk'] = df_static['eloNow'].rank(ascending=False)
    return df_static


def all_elo_df():
    active_comps = ['BL1', 'FL1', 'PL', 'ELC', 'PD', 'SA', 'PPL', 'DED']
    df_all = pd.concat([comp_elo_df(cc) for cc in active_comps], ignore_index=True)
    return df_all


df_static = comp_elo_df(initial_comp)
teams_list = [{'label': df_static.loc[ix]['name'], 'value': ix} for ix in df_static.index]
team_links = [html.Li(dcc.Link(team['label'], href=str(team['value']))) for team in teams_list]


def bkg_color_bins(df, n_bins=9, columns=['eloNow']):
    bounds = [i * (1.0 / n_bins) for i in range(n_bins + 1)]
    if columns == 'all':
        if 'id' in df:
            df_numeric_cols = df.select_dtypes('number').drop(['id'], axis=1)
        else:
            df_numeric_cols = df.select_dtypes('number')
    else:
        df_numeric_cols = df[columns]
    df_max = df_numeric_cols.max().max()
    df_min = df_numeric_cols.min().min()
    ranges = [((df_max - df_min) * i) + df_min for i in bounds]
    styles = [
        {'if': {'row_index': 'odd'}, 'backgroundColor': 'rgb(248, 248, 248)'},
        {'if': {'column_type': 'text'}, 'textAlign': 'left'},
        {'if': {'column_type': 'numeric'}, 'textAlign': 'center'},
        {'if': {'column_id': 'lastRes'}, 'textAlign': 'center'},
    ]
    legend = []
    color_scale = [
        'rgb(214,83,80)',
        'rgb(242,129,95)',
        'rgb(253,176,98)',
        'rgb(254,224,139)',
        'rgb(255,255,191)',
        'rgb(217,239,139)',
        'rgb(164,216,105)',
        'rgb(108,211,105)',
        'rgb(80,226,143)'
        ]
    for i in range(1, len(bounds)):
        min_bound = ranges[i - 1]
        max_bound = ranges[i]
        backgroundColor = color_scale[i - 1]

        for column in df_numeric_cols:
            styles.append({
                'if': {
                    'filter_query': (
                        '{{{column}}} >= {min_bound}' +
                        (' && {{{column}}} < {max_bound}' if (i < len(bounds) - 1) else '')
                    ).format(column=column, min_bound=min_bound, max_bound=max_bound),
                    'column_id': column
                },
                'backgroundColor': backgroundColor,
                'color': 'rgb(8, 8, 8)',
                'fontWeight': '500'
            })

        legend_border_left = '1px rgb(64, 64, 64) solid'
        legend_border_right = '1px rgb(64, 64, 64) solid'
        if i == 1:
            legend_border_left = '2px rgb(64, 64, 64) solid'
        elif i == n_bins:
            legend_border_right = '2px rgb(64, 64, 64) solid'
        legend.append(
            html.Div(
                style={'display': 'inline-block', 'width': '64px'},
                children=[
                    html.Div(
                        style={
                            'backgroundColor': backgroundColor,
                            'borderLeft': legend_border_left,
                            'borderRight': legend_border_right,
                            'borderTop': '2px rgb(64, 64, 64) solid',
                            'borderBottom': '2px rgb(64, 64, 64) solid',
                            'height': '32px'
                        }
                    ),
                    html.Small(round(min_bound), style={'paddingLeft': '-2px'})
                ]
            )
        )
        full_scale = html.Small(color_scale, style={'paddingLeft': '2px'})

    return (styles, html.Div(legend, style={'padding': '5px 0 5px 0'}), full_scale)


(styles, legend, full_scale) = bkg_color_bins(df_static)

elo_columns = [
    {
        'id': 'eloRk',
        'name': ['Pos.', 'Elo'],
        'type': 'numeric',
        'format': Format(
            precision=0,
            scheme=Scheme.fixed
        )
    },
    {
        'id': 'tablePos',
        'name': ['Pos.', 'Pts'],
        'type': 'numeric',
        'format': Format(
            precision=0,
            scheme=Scheme.fixed
        )
    },
    {'id': 'shortName', 'name': ['Team', 'Name'], 'type': 'text'},
    {
        'id': 'eloNow',
        'name': ['Team', 'Elo'],
        'type': 'numeric',
        'format': Format(
            precision=0,
            scheme=Scheme.fixed
        )
    },
    {
        'id': 'matches',
        'name': ['Season', 'MP'],
        'type': 'numeric',
        'format': Format(
            precision=0,
            scheme=Scheme.fixed
        )
    },
    {
        'id': 'won',
        'name': ['Season', 'W'],
        'type': 'numeric',
        'format': Format(
            precision=0,
            scheme=Scheme.fixed
        )
    },
    {
        'id': 'lost',
        'name': ['Season', 'L'],
        'type': 'numeric',
        'format': Format(
            precision=0,
            scheme=Scheme.fixed
        )
    },
    {
        'id': 'draw',
        'name': ['Season', 'D'],
        'type': 'numeric',
        'format': Format(
            precision=0,
            scheme=Scheme.fixed
        )
    },
    {
        'id': 'points',
        'name': ['Season', 'Pts'],
        'type': 'numeric',
        'format': Format(
            precision=0,
            scheme=Scheme.fixed
        )
    },
    {
        'id': 'goalsFor',
        'name': ['Goals', 'GF'],
        'type': 'numeric',
        'format': Format(
            precision=0,
            scheme=Scheme.fixed
        )
    },
    {
        'id': 'goalsAga',
        'name': ['Goals', 'GA'],
        'type': 'numeric',
        'format': Format(
            precision=0,
            scheme=Scheme.fixed
        )
    },
    {
        'id': 'goalDiff',
        'name': ['Goals', 'GD'],
        'type': 'numeric',
        'format': Format(
            precision=0,
            scheme=Scheme.fixed
        )
    },
    {
        'id': 'lastDiff',
        'name': ['Last Match', '+/-'],
        'type': 'numeric',
        'format': Format(
            precision=0,
            scheme=Scheme.fixed
        )
    },
    {'id': 'lastRes', 'name': ['Last Match', 'Res'], 'type': 'text'},
    {'id': 'lastFix', 'name': ['Last Match', 'vs.'], 'type': 'text'},
]

app = dash.Dash(
    __name__,
    assets_ignore='.*dash-default.*',
    meta_tags=[
        {
            'charset': 'utf-8'
        },
        {
            'http-equiv': 'X-UA-Compatible',
            'content': 'IE=edge'
        },
        {
            'name': 'viewport',
            'content': 'width=device-width, initial-scale=1.0'
        }
    ],
    title='Football Elo Ratings',
    update_title=None
)

layout_index = html.Div([
    dcc.Location(id='url', refresh=False),
    html.Nav(
        children=[
            html.Div(
                children=[
                    html.A(
                        "Elo Football",
                        href="/",
                        className="uk-navbar-item uk-logo"
                    ),
                    html.Ul(
                        children=[
                            html.Li(
                                html.A(
                                    "Home",
                                    href="/"
                                ),
                                className="uk-active"
                            ),
                            html.Li(
                                html.A(
                                    "PageAlt",
                                    href="/page-alt"
                                ),
                                className=""
                            ),
                            html.Li(
                                id="page-location"
                            ),
                        ],
                        className="uk-navbar-nav"
                    ),
                ],
                className="uk-navbar-left"
            ),
        ],
        className="uk-navbar-container uk-navbar-transparent"
    ),
    html.Div(
        className="uk-container uk-margin-small",
        children=[
            html.Div(
                className="uk-container uk-margin-small",
                children=[
                    html.Div(
                        className="uk-card uk-card-default uk-card-small uk-card-body",
                        children=[
                            html.Span(
                                "Choose a Competition:",
                                className="uk-text"
                            ),
                            html.Div(
                                dcc.Dropdown(
                                    id='comps-dropdown',
                                    options=df_compini.to_dict('records'),
                                    placeholder="All Competitions",
                                    value=initial_comp
                                ),
                            className='uk-margin-small'
                            )
                        ]
                    ),
                    html.Div(
                        className="uk-card uk-card-default uk-card-small uk-card-body uk-margin-small",
                        children=[
                            html.Div(
                                children=card_title,
                                className="uk-text-lead",
                                id="comp-title"
                            )
                        ]
                    )
                ],
            ),
            dcc.Tabs(
                id='tabs',
                value='league-view',
                parent_className='tabs',
                className='tabs-container',
                children=[
                    dcc.Tab(
                        label='Leagues',
                        value='league-view',
                        className='tab',
                        selected_className='tab--selected',
                        children=[
                            html.Div(
                                className="uk-card uk-card-default",
                                id="tab-content-1",
                                children=[
                                    html.Div(
                                        className="uk-card-body",
                                        children=[
                                            daq.ToggleSwitch(
                                                id='colorscale-toggle',
                                                color='#ffffff',
                                                label='Show color scale',
                                                size=36,
                                                style={'margin-bottom': '1.0em'},
                                                value=False
                                            ),
                                            dash_table.DataTable(
                                                id='elolgtable',
                                                columns=elo_columns,
                                                data=df_static.to_dict('records'),
                                                sort_action='custom',
                                                sort_mode='single',
                                                sort_by=[],
                                                style_cell={
                                                    'fontFamily': 'Inter var, Inter, Roboto, Nunito, Arial, sans-serif',
                                                    'fontSize': '17px'
                                                },
                                                style_cell_conditional=[
                                                    {'if': {'column_id': 'eloRk'},
                                                    'width': '5%'},
                                                    {'if': {'column_id': 'tablePos'},
                                                    'width': '5%'},
                                                    {'if': {'column_id': 'shortName'},
                                                    'width': '20%'},
                                                    {'if': {'column_id': 'eloNow'},
                                                    'width': '10%'},
                                                    {'if': {'column_id': 'matches'},
                                                    'width': '5%'},
                                                    {'if': {'column_id': 'won'},
                                                    'width': '5%'},
                                                    {'if': {'column_id': 'lost'},
                                                    'width': '5%'},
                                                    {'if': {'column_id': 'draw'},
                                                    'width': '5%'},
                                                    {'if': {'column_id': 'points'},
                                                    'width': '5%'},
                                                    {'if': {'column_id': 'goalsFor'},
                                                    'width': '5%'},
                                                    {'if': {'column_id': 'goalsAga'},
                                                    'width': '5%'},
                                                    {'if': {'column_id': 'goalDiff'},
                                                    'width': '5%'},
                                                    {'if': {'column_id': 'lastDiff'},
                                                    'width': '5%'},
                                                    {'if': {'column_id': 'lastRes'},
                                                    'width': '5%'},
                                                    {'if': {'column_id': 'lastFix'},
                                                    'width': '5%'},
                                                ],
                                                style_data_conditional=styles,
                                                style_header={
                                                    'backgroundColor': 'rgb(69, 74, 79)',
                                                    'color':  'rgb(233, 233, 240)',
                                                    'fontFamily': 'Inter, Roboto, Nunito, Arial, sans-serif',
                                                    'fontSize': '18px',
                                                    'fontWeight': '400',
                                                    'paddingLeft': '8px',
                                                    'paddingRight': '8px',
                                                },
                                                style_header_conditional=[
                                                    {'if': {'column_type': 'text'}, 'textAlign': 'left'},
                                                    {'if': {'column_type': 'numeric'}, 'textAlign': 'center'},
                                                ],
                                                merge_duplicate_headers=True,
                                                style_as_list_view=False,
                                            ),
                                            html.Div(
                                                className="uk-card uk-card-small uk-padding-remove uk-margin-small uk-align-right",
                                                children=[
                                                    html.Div(
                                                        children=legend,
                                                        id="colorbins-legend",
                                                        className="uk-card-body"
                                                    )
                                                ]
                                            )
                                        ]
                                    )
                                ]
                            )
                        ]
                    ),
                    dcc.Tab(
                        label='Teams',
                        value='team-view',
                        className='tab',
                        selected_className='tab--selected',
                        children=[
                            html.Div(
                                className="uk-card uk-card-default",
                                id="tab-content-2",
                                children=[
                                    html.Div(
                                        className="uk-card-body",
                                        children=[
                                            html.Div([
                                                html.P(
                                                    "Teams List",
                                                    className="uk-text-lead"
                                                ),
                                            ],
                                            className="uk-margin-small"
                                            ),
                                            html.Ul(
                                                id="team-links",
                                                className="uk-list",
                                                children=team_links
                                            )
                                        ],
                                    ),
                                ],
                            ),
                        ],
                    ),
                ],
            ),
            html.Div(
                html.Div(
                    children=[
                        html.P(
                            children=[
                                "Created by Alex Munger – 2020 | ",
                                html.A(
                                    "GitHub",
                                    href="https://amunger3.github.io"
                                ),
                            ],
                            className="uk-text-center"
                        ),
                    ],
                    className="uk-container"
                ),
                className="uk-margin-medium"
            ),
        ]
    )
])

server = app.server
app.layout = layout_index


# Callbacks
@app.callback(dash.dependencies.Output('page-location', 'children'),
              [dash.dependencies.Input('url', 'pathname')])
def display_page(pathname):
    return html.A(pathname, href="#")


@app.callback(
    [Output('elolgtable', 'data'),
    Output('elolgtable', 'style_data_conditional'),
    Output('comp-title', 'children'),
    Output('colorbins-legend', 'children'),
    Output('team-links', 'children')],
    [Input('comps-dropdown', 'value'),
    Input('elolgtable', 'sort_by'),
    Input('colorscale-toggle', 'value')])
def update_table(value, sort_by, colorbins_bool):
    if value:
        df_static = comp_elo_df(value)
        card_title = set_table_header(value)
    else:
        df_static = all_elo_df()
        card_title = 'All Competitions'

    if len(sort_by):
        df_sorted = df_static.sort_values(
            sort_by[0]['column_id'],
            ascending=sort_by[0]['direction'] == 'asc'
        )
    else:
        df_sorted = df_static.sort_values(
            'eloNow',
            ascending=False
        )
    teams_list = [{'label': df_static.loc[ix]['name'], 'value': ix} for ix in df_static.index]
    team_links = [html.Li(dcc.Link(team['label'], href=str(team['value']))) for team in teams_list]
    (styles, legend, full_scale) = bkg_color_bins(df_static)

    return df_sorted.to_dict('records'), styles, card_title, legend, team_links


if __name__ == '__main__':
    app.run_server(debug=True)
